integration-tests/test/test_slash.py

```diff
@@ -155,11 +157,94 @@ def test_slash_justification_not_correct(command_line_options: CommandLineOption
+            Justification(validator=BOOTSTRAP_NODE_KEY.get_public_key().to_bytes(), latestBlockHash=bytes.fromhex(genesis_block['blockHash'])),
+        ])
+        client.send_block(invalid_block, validator2)
+
+        wait_for_node_sees_block(context, validator2, evil_block_hash.hex())
+
+        # block which is created by validator2 but not slashing validator1
+        block_not_slash_invalid_block = BlockMessage()
+        block_not_slash_invalid_block.CopyFrom(block_msg)
+        block_not_slash_invalid_block.seqNum = 1
+        block_not_slash_invalid_block.body.state.blockNumber = 1  # pylint: disable=maybe-no-member
+        block_not_slash_invalid_block.sender = BONDED_VALIDATOR_KEY_2.get_public_key().to_bytes()
+        block_not_slash_invalid_block.ClearField("justifications")
+        block_not_slash_invalid_block.justifications.extend([  # pylint: disable=maybe-no-member
+            Justification(validator=BONDED_VALIDATOR_KEY_1.get_public_key().to_bytes(), latestBlockHash=evil_block_hash),
+            Justification(validator=BONDED_VALIDATOR_KEY_2.get_public_key().to_bytes(), latestBlockHash=bytes.fromhex(genesis_block['blockHash'])),
+            Justification(validator=BOOTSTRAP_NODE_KEY.get_public_key().to_bytes(), latestBlockHash=bytes.fromhex(genesis_block['blockHash'])),
+        ])
+        deploy_data = create_deploy_data(BONDED_VALIDATOR_KEY_2, Path("../rholang/examples/tut-hello.rho").read_text(), 1, 1000000)
+        block_not_slash_invalid_block.body.deploys[0].deploy.CopyFrom(deploy_data)  # pylint: disable=maybe-no-member
+        block_not_slash_invalid_block.header.ClearField("parentsHashList")  # pylint: disable=maybe-no-member
+        block_not_slash_invalid_block.header.parentsHashList.append(bytes.fromhex(genesis_block['blockHash']))  # pylint: disable=maybe-no-member
+        block_not_slash_invalid_block.header.timestamp = int(time.time()*1000)  # pylint: disable=maybe-no-member
+        block_not_slash_invalid_block.header.deploysHash = gen_deploys_hash_from_block(block_not_slash_invalid_block)  # pylint: disable=maybe-no-member
+        invalid_block_hash = gen_block_hash_from_block(block_not_slash_invalid_block)
+        block_not_slash_invalid_block.sig = BONDED_VALIDATOR_KEY_2.sign_block_hash(invalid_block_hash)
+        block_not_slash_invalid_block.blockHash = invalid_block_hash
+
+        client.send_block(block_not_slash_invalid_block, validator3)
+
+        # Because validator2 doesn't slash validator1's block while validator3 slash validator1's block,
+        # hence there are some comm events lack of in validator3 which cause an invalid transaction
+        record_invalid = re.compile("Recording invalid block {}... for InvalidTransaction".format(invalid_block_hash.hex()[:10]))
+        wait_for_log_match(context, validator3, record_invalid)
+
+
+        validator3.deploy(contract, BOOTSTRAP_NODE_KEY)
```

Comment:
Do you use `BOOTSTRAP_NODE_KEY` here just as a _random_ deployer key?

Reply:
> So this does the following?
> 
> 1. v1 proposes valid block
> 2. v1 creates another block with invalid parent block hash and sends it to v2
> 3. v2 creates block using v1's second block as parent hash (i.e. no slashing, another invalid block) and sends it to v3
> 4. v3 records invalid block (`InvalidTransaction`) from v2
> 5. v3 proposes block which slashes both v1 and v2

You are right!



> Do you use `BOOTSTRAP_NODE_KEY` here just as a _random_ deployer key?

Actually it doesn't matter which key as long as validator3 proposed a new block.

Reply:
```v1 proposes valid block
v1 creates another block with invalid parent block hash and sends it to v2
v2 creates block using v1's second block as parent hash (i.e. no slashing, another invalid block) and sends it to v3
v3 records invalid block (InvalidTransaction) from v2
v3 proposes block which slashes both v1 and v2```

Can we have this put into the docstring?  This is a great summary of what the test scenario does.

Reply:
Thanks 👍